File: ios/calorietracker/Actions/ActionDeepLink.swift
import Foundation

/// `ayuvo://action/<id>?k=v` and `ayuvo://open/<section>` — Swift port of `parse_deeplink` in
/// `scripts/actions_reference.py` (vectors: `shared/actions/test-vectors/deeplinks.json`). Parsed by hand
/// on the raw string: `URLComponents` does not turn `+` into a space and normalises some escapes.
nonisolated enum ActionDeepLink {
    nonisolated struct Request: Equatable, Sendable {
        let id: String
        let params: [String: String]
    }

    nonisolated enum Parsed: Equatable, Sendable {
        case request(Request)
        /// Not ours (widgets, log-food, share hand-offs…): fall through to the other handlers.
        case notActionLink
        case failure(code: String)
    }

    static func parse(_ url: String) -> Parsed {
        let prefix = "ayuvo://"
        guard url.hasPrefix(prefix) else { return .notActionLink }
        var rest = Substring(url.dropFirst(prefix.count))
        if let hash = rest.firstIndex(of: "#") { rest = rest[..<hash] }
        let path: Substring, query: Substring
        if let mark = rest.firstIndex(of: "?") {
            path = rest[..<mark]
            query = rest[rest.index(after: mark)...]
        } else {
            path = rest
            query = ""
        }
        let host: Substring, tailRaw: Substring
        if let slash = path.firstIndex(of: "/") {
            host = path[..<slash]
            tailRaw = path[path.index(after: slash)...]
        } else {
            host = path
            tailRaw = ""
        }
        guard host == "action" || host == "open" else { return .notActionLink }
        var tail = tailRaw
        if tail.hasSuffix("/") { tail = tail.dropLast() }
        guard !tail.isEmpty, !tail.contains("/") else { return .failure(code: "bad_link") }
        guard let segment = percentDecode(tail), !segment.isEmpty else { return .failure(code: "bad_link") }
        var params: [String: String] = [:]
        if !query.isEmpty {
            for piece in query.split(separator: "&", omittingEmptySubsequences: false) where !piece.isEmpty {
                let key: Substring, value: Substring
                if let eq = piece.firstIndex(of: "=") {
                    key = piece[..<eq]
                    value = piece[piece.index(after: eq)...]
                } else {
                    key = piece
                    value = ""
                }
                guard let decodedKey = percentDecode(key), !decodedKey.isEmpty,
                      let decodedValue = percentDecode(value)
                else { return .failure(code: "bad_link") }
                if params[decodedKey] != nil { return .failure(code: "duplicate_param") }
                params[decodedKey] = decodedValue
            }
        }
        if host == "open" { return .request(Request(id: "open.section", params: ["section": segment])) }
        return .request(Request(id: segment, params: params))
    }

    /// Percent-decode with `+` as space; nil when malformed or not UTF-8.
    static func percentDecode<S: StringProtocol>(_ text: S) -> String? {
        var bytes: [UInt8] = []
        let scalars = Array(text.unicodeScalars)
        var i = 0
        while i < scalars.count {
            let c = scalars[i]
            if c == "%" {
                guard i + 2 < scalars.count, let hi = hexValue(scalars[i + 1]), let lo = hexValue(scalars[i + 2]) else { return nil }
                bytes.append(UInt8(hi * 16 + lo))
                i += 3
                continue
            }
            if c == "+" {
                bytes.append(0x20)
            } else {
                bytes.append(contentsOf: Array(String(c).utf8))
            }
            i += 1
        }
        return String(data: Data(bytes), encoding: .utf8)
    }

    private static func hexValue(_ c: Unicode.Scalar) -> Int? {
        switch c.value {
        case 0x30...0x39: return Int(c.value - 0x30)
        case 0x41...0x46: return Int(c.value - 0x41 + 10)
        case 0x61...0x66: return Int(c.value - 0x61 + 10)
        default: return nil
        }
    }
}

File: android/app/src/main/java/com/ayuvo/health/actions/ActionDeepLink.kt
package com.ayuvo.health.actions

import java.io.ByteArrayOutputStream
import java.nio.ByteBuffer
import java.nio.charset.CharacterCodingException
import java.nio.charset.CodingErrorAction

/**
 * `ayuvo://action/<id>?k=v` and `ayuvo://open/<section>` (docs/actions.md), ported from
 * `parse_deeplink` in scripts/actions_reference.py. Parsed by hand (not `Uri.getQueryParameter`)
 * so '+' and percent-escapes decode exactly as on iOS. Other `ayuvo://` hosts (widget links) are
 * [LinkError.NOT_ACTION_LINK] and keep their existing handlers.
 */
object ActionDeepLink {
    enum class LinkError(val raw: String) { NOT_ACTION_LINK("not_action_link"), BAD_LINK("bad_link"), DUPLICATE_PARAM("duplicate_param") }

    sealed interface Parsed {
        data class Ok(val id: String, val params: Map<String, String>) : Parsed
        data class Failed(val error: LinkError) : Parsed
    }

    private const val PREFIX = "ayuvo://"

    fun isActionLink(url: String?): Boolean =
        url != null && (url.startsWith("${PREFIX}action/") || url.startsWith("${PREFIX}open/"))

    fun parse(url: String): Parsed {
        if (!url.startsWith(PREFIX)) return Parsed.Failed(LinkError.NOT_ACTION_LINK)
        val rest = url.substring(PREFIX.length).substringBefore('#')
        val path = rest.substringBefore('?')
        val query = if ('?' in rest) rest.substringAfter('?') else ""
        val host = path.substringBefore('/')
        var tail = if ('/' in path) path.substringAfter('/') else ""
        if (host != "action" && host != "open") return Parsed.Failed(LinkError.NOT_ACTION_LINK)
        if (tail.endsWith("/")) tail = tail.dropLast(1)
        if (tail.isEmpty() || '/' in tail) return Parsed.Failed(LinkError.BAD_LINK)
        val segment = percentDecode(tail)
        if (segment.isNullOrEmpty()) return Parsed.Failed(LinkError.BAD_LINK)
        val params = LinkedHashMap<String, String>()
        if (query.isNotEmpty()) {
            for (piece in query.split('&')) {
                if (piece.isEmpty()) continue
                val k = piece.substringBefore('=')
                val v = if ('=' in piece) piece.substringAfter('=') else ""
                val key = percentDecode(k)
                val value = percentDecode(v)
                if (key.isNullOrEmpty() || value == null) return Parsed.Failed(LinkError.BAD_LINK)
                if (key in params) return Parsed.Failed(LinkError.DUPLICATE_PARAM)
                params[key] = value
            }
        }
        if (host == "open") return Parsed.Ok("open.section", mapOf("section" to segment))
        return Parsed.Ok(segment, params)
    }

    private fun hex(c: Char): Int? = when (c) {
        in '0'..'9' -> c - '0'
        in 'a'..'f' -> c - 'a' + 10
        in 'A'..'F' -> c - 'A' + 10
        else -> null
    }

    /** Percent-decoding with '+' as space; null when an escape is malformed or the bytes are not UTF-8. */
    fun percentDecode(s: String): String? {
        val out = ByteArrayOutputStream()
        var i = 0
        while (i < s.length) {
            val c = s[i]
            if (c == '%') {
                val h = s.getOrNull(i + 1)?.let(::hex) ?: return null
                val l = s.getOrNull(i + 2)?.let(::hex) ?: return null
                out.write(h * 16 + l)
                i += 3
                continue
            }
            if (c == '+') {
                out.write(0x20)
            } else {
                val cp = s.codePointAt(i)
                out.write(String(Character.toChars(cp)).toByteArray(Charsets.UTF_8))
                i += Character.charCount(cp)
                continue
            }
            i++
        }
        return try {
            Charsets.UTF_8.newDecoder()
                .onMalformedInput(CodingErrorAction.REPORT)
                .onUnmappableCharacter(CodingErrorAction.REPORT)
                .decode(ByteBuffer.wrap(out.toByteArray()))
                .toString()
        } catch (_: CharacterCodingException) {
            null
        }
    }
}
